Add support for PCI BAR reprogramming

sboeuf opened the pull request.

sboeuf commented:
@rbradford @sameo PTAL

sboeuf commented:
@rbradford @sameo 
I'm stuck with this PR and I need help if you have some spare time.
Basically, it happens that reboot does not work because the `Drop` trait is not invoked for `VirtioPciDevice` and the associated virtio device. I identified that it is failing because I have `io_bus: Arc<devices::Bus>` or `mmio_bus: Arc<devices::Bus>` embedded into my structure `AddressManager`. The structure `AddressManager` is the one implementing the `VirtioPciBarReprogramming` and `VfioPciBarReprogramming` traits.
I have spent some time looking at `vmm/src/vm.rs` and `vmm/src/lib.rs` but I don't understand how the `io_bus` and `mmio_bus` are preventing the Drop trait from being invoked.
I can see the VCPU thread being properly terminated, and the `DeviceManager` being dropped, so it comes down to the `VirtioPciBarReprog` instance which still holds the `VirtioPciBarReprogramming` trait implementation.

Let me know if you can find the reason why this is failing, or if you have some leads!

rbradford commented on vmm/src/device_manager.rs:
```diff
@@ -1250,12 +1379,12 @@ impl DeviceManager {
 
-    pub fn io_bus(&self) -> &devices::Bus {
-        &self.io_bus
+    pub fn io_bus(&self) -> &Arc<devices::Bus> {
```
I'd rather we kept ownership on DeviceManager rather than moving the ownership to the AddressManager, it feels a bit "weird". This shouldn't cause any problems as it's an Arc<> so should it can just take a clone as part of its contruction

rbradford commented:
I did a bisect and identified the "bad" commit as:

commit 10b4674fba972ffc1c4c081f6340e12521082e85 (refs/bisect/bad)
Author: Sebastien Boeuf <sebastien.boeuf@intel.com>
Date:   Wed Oct 23 17:57:26 2019 -0700

    vmm: Enable virtio-pci BAR reprogramming
    
    Signed-off-by: Sebastien Boeuf <sebastien.boeuf@intel.com>

sboeuf commented:
@rbradford yes that's the bad commit, since this is the one effectively passing the `AddressManager` which owns both `io_bus` and `mmio_bus` to the VirtioPciDevice.

rbradford commented:
I've pushed a change to switch to a weak reference in some places. Lets see if it passes the build.

sboeuf commented:
After offline discussions with @rbradford here is a summary of what needs to be done to make the implementation a bit more straightforward (this will remove one layer of traits).
- Have only one trait called `DeviceRelocation` defined by the `pci` crate.
- Pass the `DeviceRelocation` trait to `PciBus`
- `AddressManager` will implement `DeviceRelocation`
- `AddressManager` will have weak references onto `io_bus` and `mmio_bus` while `DeviceManager` will have the strong references.
- Detect BAR reprogramming from `PciConfigIo` and `PciConfigMmio`
- Add `move_bar()` to the `PciDevice` trait
- When BAR reprogramming happens, use `PciBus` to call into `DeviceRelocation` implementation of `move_at()`, and then call into the associated `PciDevice` implementation of `move_bar()`.

sboeuf commented:
@rbradford @sameo 
Ok I reworked the PR mostly as agreed with @rbradford and as described [here](https://github.com/intel/cloud-hypervisor/pull/385#issuecomment-546442775).
I fixed the CI and extended it with a dedicated test.
You can proceed with the review, let me know what you think about it :)
